src/compute/src/render/reduce.rs

```diff
@@ -644,7 +666,7 @@ where
                 // Negative counts would be surprising, but until we are 100% certain we wont
                 // see them, we should report when we do. We may want to bake even more info
                 // in here in the future.
-                if source.iter().any(|(_val, cnt)| cnt < &0) {
+                if validating && source.iter().any(|(_val, cnt)| cnt < &0) {
```

Comment:
I might be confused, but isn't `validating` always `false` here?

* We set it to `true` in the beginning.
* The first call to `build_bucketed_stage` receives the `true` value, which causes it to output `Some` error stream.
* Because we have `Some` error stream, we enter the `if let` on line 650 and set `validating = false`.
* `validating` remains `false` until we get here.

So the only way it can be `true` here is we build zero stages, which probably can't happen?

But by the same logic, won't the `err_output` we return from this function always be `Some` in the end too?

Reply:
It can happen that we build zero stages. Consider the logic where we compute the buckets: https://github.com/MaterializeInc/materialize/blob/3344fd693c57843770a322a6e58b4f8c70511cb5/src/compute-client/src/plan/reduce.rs#L670-L688
If `expected_group_size` is given as 16 or smaller, then the `buckets` vector ends up empty. The expected group size parameter can be set as a query hint: https://materialize.com/docs/sql/select/#using-query-hints.
I think it is by design that a user can tune down the memory requirements here to a single arrangement of the input without a hierarchy, provided that the user has some domain-specific knowledge about cardinalities.

Reply:
Makes sense. And you point that out in the PR description too, sorry for not catching this!

Reply:
No worries, thanks for double-checking!